### ISSUE TITLE
In-app exporter only returning one image

### ISSUE THREAD

**Issue by jywarren** [publiclab/mapknitter#940]
Strangely, similarly to #919, the in-app exporter seems to be running on only one image. Investigating; https://mapknitter.org/maps/nottingham-city-centre/

**Comment by jywarren** [comment]
I'm going to test on https://mapknitter.org/maps/ceres--2/ for simplicity.

**Comment by jywarren** [comment]
OK, `ceres--2` is also showing just one image in the export. So, reproduced. I'm re-running the export, and it says `warping 1 of 3`, so i know it can see the images. and is running through them. On completion, it's still just one image. 

...now to know why it isn't working... let me check the exporter gem running separately! 

http://export.mapknitter.org/export?url=https://mapknitter.org/maps/ceres--2/warpables.json&scale=30 is returning http://export.mapknitter.org/id/1565815972/status.json -- it'll take a while...

**Comment by jywarren** [comment]
The exporter gem in the cloud app can correctly generate mutiple image exports, I just confirmed this by running it manually at export.mapknitter.org. 

Now trying to run this at http://mapknitter-stable.laboratoriopublico.org/maps/ceres--2 for replication!

**Comment by jywarren** [comment]
I'm manually downloading all 3 individual pre-composite GeoTiffs from the mapknitter.org ceres export i just ran.

**Comment by jywarren** [comment]
(noting that staging export also just ran and only produced 1 image)

**Comment by jywarren** [comment]
OK, and confirming that all three GeoTiffs were good...

**Comment by jywarren** [comment]
I'm running the export in the console now on the production server to try to track what happens...

**Comment by jywarren** [comment]
OK, I believe this is the command and output; noting that it again produced just one image in the composite:

```
gdalwarp -s_srs EPSG:3857 -te 144.9828654528 -37.7668304285 144.9859553576 -37.7650239004 public/warps/ceres--2/306187-geo.tif public/warps/ceres--2/ceres--2-geo.tif
Creating output file that is 1272P x 744L.
Processing input file public/warps/ceres--2/306187-geo.tif.
Using band 4 of source image as alpha.
Using band 4 of destination image as alpha.
0...10...20...30...40...50...60...70...80...90...100 - done.
Warning 1: Metadata exceeding 32000 bytes cannot be written into GeoTIFF. Transferred to PAM instead.
gdalwarp public/warps/ceres--2/306188-geo.tif public/warps/ceres--2/ceres--2-geo.tif
Processing input file public/warps/ceres--2/306188-geo.tif.
Using band 4 of source image as alpha.
Using band 4 of destination image as alpha.
0...10...20...30...40...50...60...70...80...90...100 - done.
Warning 1: Metadata exceeding 32000 bytes cannot be written into GeoTIFF. Transferred to PAM instead.
gdalwarp public/warps/ceres--2/306189-geo.tif public/warps/ceres--2/ceres--2-geo.tif
Processing input file public/warps/ceres--2/306189-geo.tif.
Using band 4 of source image as alpha.
Using band 4 of destination image as alpha.
0...10...20...30...40...50...60...70...80...90...100 - done.
Warning 1: Metadata exceeding 32000 bytes cannot be written into GeoTIFF. Transferred to PAM instead.
3.804MB
1272
744
```

**Comment by jywarren** [comment]
Seems to match these lines:

https://github.com/publiclab/mapknitter/blob/f77bef70f8afa8249de0aa3ae8907998d38f198b/lib/exporter.rb#L262-L269

**Comment by jywarren** [comment]
I manually went through these steps:

```
mapknitter@tycho2:~/production$ gdalwarp public/warps/ceres--2/306188-geo.tif public/warps/ceres--2/ceres--2-geo.tif
Processing input file public/warps/ceres--2/306188-geo.tif.
Using band 4 of source image as alpha.
Using band 4 of destination image as alpha.
0...10...20...30...40...50...60...70...80...90...100 - done.
Warning 1: Metadata exceeding 32000 bytes cannot be written into GeoTIFF. Transferred to PAM instead.
mapknitter@tycho2:~/production$ dalwarp public/warps/ceres--2/306189-geo.tif public/warps/ceres--2/ceres--2-geo.tif
-su: dalwarp: command not found
mapknitter@tycho2:~/production$ gdalwarp public/warps/ceres--2/306189-geo.tif public/warps/ceres--2/ceres--2-geo.tif
Processing input file public/warps/ceres--2/306189-geo.tif.
Using band 4 of source image as alpha.
Using band 4 of destination image as alpha.
0...10...20...30...40...50...60...70...80...90...100 - done.
Warning 1: Metadata exceeding 32000 bytes cannot be written into GeoTIFF. Transferred to PAM instead.
mapknitter@tycho2:~/production$ convert -background white -flatten public/warps/ceres--2/ceres--2-geo.tif public/warps/ceres--2/ceres--2.jpg           convert-im6.q16: Unknown field with tag 33550 (0x830e) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
convert-im6.q16: Unknown field with tag 33922 (0x8482) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
convert-im6.q16: Unknown field with tag 34735 (0x87af) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
convert-im6.q16: Unknown field with tag 34737 (0x87b1) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/912.
```

And still got one image. Are we losing critical metadata in the PAM or something?

**Comment by jywarren** [comment]
OK, noting that in the exporter gem, we're no longer distinguishing between "first" warpable and others, despite the fact that we initialize the `first` variable here: 

https://github.com/publiclab/mapknitter-exporter/blob/c75c31d43b0d834c8198d43f857f4483ed5ce534/lib/mapknitterExporter.rb#L268

I'm going to try manually running the commands from that version instead.

**Comment by jywarren** [comment]
That worked. Refactoring now.

**Comment by jywarren** [comment]
Fix at c76748ec33056db9baccfca0717e360e51a2bc74

**Comment by jywarren** [comment]
Published to production and testing at https://mapknitter.org/maps/ceres--2/

**Comment by jywarren** [comment]
Hmm. Must've made a mistake...

```
gdalwarp -s_srs EPSG:3857 -te 144.9828654528 -37.7668304285 144.9859553576 -37.7650239004  public/warps/ceres--2/306187-geo.tif public/warps/ceres--2/306188-geo.tif public/warps/ceres--2/306189-geo.tif public/warps/ceres--2/ceres--2.tif 
Creating output file that is 1908P x 1116L.
Processing input file public/warps/ceres--2/306187-geo.tif.
Using band 4 of source image as alpha.
Using band 4 of destination image as alpha. 
0...10...20...30...40...50Warning 1: Metadata exceeding 32000 bytes cannot be written into GeoTIFF. Transferred to PAM instead.
...60...70...80...90...100 - done.
Processing input file public/warps/ceres--2/306188-geo.tif.
Using band 4 of source image as alpha.
Using band 4 of destination image as alpha.
0...10...20...30...40...50...60...70...80...90...100 - done.
Processing input file public/warps/ceres--2/306189-geo.tif.
Using band 4 of source image as alpha.
Using band 4 of destination image as alpha.
0...10...20...30...40...50...60...70...80...90...100 - done.
identify-im6.q16: unable to open image `public/warps/ceres--2/ceres--2-geo.tif': No such file or directory @ error/blob.c/OpenBlob/2701.
```

**Comment by jywarren** [comment]
I forgot `-geo`. Final fixes in c615554471c50a418223496c537a32aef1dafb06 and 619fcc43759a6a0bfbdeb21af7a6ad44386914a8. Closing!
